### ISSUE TITLE
Proposal: type promotion logic (torch.result_type)

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#9515]
This proposes an algorithm for computing the result type of a mixed-type operation like `torch.add`. The function will be exposed to Python via `torch.result_type(*tensors)`. This proposal covers the default result type calculation; some operators may override the default behavior.

The similar NumPy function is [`numpy.result_type`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.result_type.html).

**Wrapped numbers**

A Tensor is a considered a "wrapped number" if it is auto-wrapped from a C++ or Python number type. Integer types are wrapped as 0-dim int64 tensors and floating-point types are wrapped as 0-dim double tensors. All wrapped numbers are 0-dim tensors, but not all 0-dim tensors are wrapped numbers. In general, wrapped numbers behave like normal 0-dim tensors, except they are handled specially in the `torch.result_type` calculation.

For example, in `tensor + 5` and `torch.add(tensor, 5)`, `5` gets wrapped as a 0-dim `torch.int64` (a "wrapped number").  However, `torch.add(tensor, torch.tensor(5))` does not have a wrapped number because `torch.tensor(5)` is an explicit construction. Wrapped number status does not propagate to returned tensors. Returned tensors are never considered wrapped numbers.

**Result type calculation**

Each operand has a category (integer or floating-point) and a priority:

1) Tensors of dimension 1 or larger
2) Tensors of dimension 0 that are not wrapped numbers
3) Wrapped numbers

By default, only the highest priority operands participate in the type promotion logic. Lower priority operands participate if their category (e.g. floating-point) is of higher rank than any higher priority operands (e.g. integers).

In pseudo-code the result-type calculation is:

```
def result_type(*args):
  return promote_types(infer_scalar_type(arg) for arg in args if participates(arg, args))

def infer_scalar_type(arg):
  if is_wrapped_number(arg):
    return torch.get_default_dtype() if is_floating_point(arg) else torch.int64
  else:
    return arg.dtype

def participates(arg, args):
  if priority(arg) >= max(priority(other) for other in args):
    return True
  if category(arg) > max(category(other) for other in args if priority(other) > priority(arg)):
   return True
  return False

def priority(arg):
  if arg.dim() > 0: return 3
  elif not is_wrapped_number(arg): return 2
  else: return 1

def category(arg):
  if is_floating_point(arg): return 2
  else: return 1
```

Examples (assuming default float32 tensor dtype):

```
randn(3, dtype=float32) * 5 -> float32
tensor([0, 0, 1], dtype=uint8) + 1 -> uint8
tensor([0, 0, 1], dtype=uint8) + 1000 -> uint8  # NOTE: integer overflow
tensor([0, 0, 1], dtype=uint8) + 5.5 -> float32 (default tensor dtype)
tensor([0, 0, 1], dtype=uint8) + tensor(5.5, dtype=double) -> double

randn(3, dtype=float32) + tensor(5.5, dtype=double) -> float32
tensor(5.5, dtype=float16) + 2.2 -> float16
tensor(5.5, dtype=float16) + 100000 -> float16 # NOTE: inf
tensor(5.5, dtype=float16) + tensor(100000.0) -> float32 (default tensor dtype)
```

Appendix:

**Why don't we use NumPy's behavior?**

NumPy's result_type logic has two undesirable behaviors. The first is that requires examining the actual value of scalars (and 0-dim arrays). This would require a host-device synchronization for 0-dim CUDA tensors. The second is that it often up-promotes 0-dim arrays to float64 or int64. For example:

```
type(np.array(4.0, dtype=np.float32) + 1) -> np.float64
type(np.array(0, dtype=np.uint8) + 1) -> np.int64
```

**Comment by vadimkantorov** [comment]
How about ˋtorch.FloatTensor([0.1, 0.2]) * torch.ByteTensor([0, 1])`? (very frequent when dealing with masks)

**Comment by tugrulates** [comment]
Extending the proposal a bit more to address the following:

- *In-place & out functions:* How to make sure to not lose data silently?
- *Unary operations:* How to `sin(int)` is a `float`?
- *Special params/op:* How to take function semantics into account (e.g. boolean for masking params)?

**TL;DR** NumPy style `casting` API

### 1. Casting API

This section defining the default behavior for most operations. Exemptions are discussed in the next section.

#### 1.1. Function parameters for casting

Operations will take additional parameters to let the user have control in casting behavior.

`torch.pow(input, exponent, casting='safe', dtype=torch.float32)`

**`casting`:** Lets the user determine their comfort level with casting. The values it can take are:
- `safe` (default): Execute if output type can preserve all values of result type, regardless of actual Tensor values
- `unsafe`: Execute if any casting can be done
- `no`: Execute only if all no casting would take place

**`dtype`:** Lets the user determine the final type, in absence of an implied type from the output Tensor (in-place and out operators).

#### 1.2. Output type vs. result type

*Result type* is the type that is inferred from operators taking place in the operation determined by `torch.result_type(inputs_and_outputs)`. All inputs and the output can always be safely promoted to the result type.

*Output type* is type of the data used inside the inner loop and the generated output. Output type will be determined with the following waterfall:

- Type of operating tensor for in-place operations
- Type of `out` parameter if given
- `dtype` parameter if given
- result type

The operation will not execute if casting is not allowed from result type to output type. This will be controlled by `torch.can_cast(from, to, casting)`, where `casting` is API user's choice of cast safety.

This is similar to NumPy's [np.can_cast](https://docs.scipy.org/doc/numpy/reference/generated/numpy.can_cast.html).

#### Example:

*Assuming promote_types(int32, float32) -> float64, which is the NumPy behavior.*

```
>>> a = torch.ones(4, dtype=torch.float32)
>>> b = torch.ones(4, dtype=torch.int32)
>>> torch.pow(a, b).dtype
torch.float64
>>> torch.pow(a, b, casting='no').dtype
TypeError: IntTensor can't be coerced into DoubleTensor due to casting rule 'no'
>>> torch.pow(a, b, dtype=torch.float32).dtype
TypeError: DoubleTensor can't be coerced into FloatTensor due to casting rule 'safe'
>>> torch.pow(a, b, dtype=torch.float32, casting='unsafe').dtype
torch.float32
>>> torch.pow(a, b, out=a, casting='unsafe').dtype
torch.float32
>>> a.pow_(b).dtype
TypeError: DoubleTensor can't be coerced into FloatTensor due to casting rule 'safe'
>>> a.pow_(b, casting='unsafe').dtype
torch.float32
```

### 1.3. Type conversion API changes

- New param `casting` for [`Tensor.type_as`](https://pytorch.org/docs/stable/tensors.html?highlight=type_as#torch.Tensor.type_as). This will start enforcing casting safety for converting types but the default value will default to `unsafe`.
- New param `casting` for [`torch.as_tensor`](https://pytorch.org/docs/stable/torch.html?highlight=ones_like#torch.as_tensor), with same behavior and default.

### 2. Implementation for native functions

#### 2.1. Multi-type casting behavior

Native functions will set a casting behavior using the `casting` attribute. For example:

```
- func: pow(Tensor self, Tensor exponent) -> Tensor
  casting: auto
```

Values this can take are:
`auto` (default): Handle casting at dispatch, creating copies of input where necessary
`no`: casting is exposed at the API layer (e.g. ones_like)
`native`: Native function handles casting itself due to one of:
- a more optimized solution is available at the loop level
- semantics of the function require different casting rules (e.g. masking)

Long term plan is to move most functions to `native` and provide the default casting implementation from `TensorIterator`. However, proposing to have `auto` as default in order to separate these two lines of work.

Since, for an operation, one backend can have optimized multi-type loops but the others might not, it would be possible to customize by backend:

```
- func: pow(Tensor self, Tensor exponent) -> Tensor
  casting:
    CPU: auto
    CUDA: native
```

#### 2.2. Special case for single type casting

Most native functions are currently only implemented for floating types and they are missing implementations for integral types.

```
>>> torch.sin(torch.tensor(4))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "sin" not implemented for 'torch.LongTensor'
```

The casting API can solve this by taking into account what types the function supports natively.

```
- func: pow(Tensor self, Tensor exponent) -> Tensor
  types: Float, Double
```

If the native function explicitly states that it supports a confined list of types, the types will be used in `result_type` calculation in order until one that can be casted into is found. Hence, this would always allow operating with lower types than the native function supports.

NumPy's behavior is slightly different here:
- Every ufunc defines the types it can operate on
- An unsafe downcast is possible if the ufunc does not have loops for higher types

The difference is accepted for simplification since PyTorch typeset is a lot more confined, i.e. almost all functions support the highest kind, floating points.

### 3. Notes

- Proposed behavior applies to cpp extension as well.
- Omitting `equiv` casting from NumPy because PyTorch always works on native endianness.
- Omitting `same_kind` casting from NumPy because `torch.dtype.kind` isn't a thing.
- This won't handle complex types for now but it is flexible enough to extend (native casting and introduce `dtype.kind`)
- Native operations can still continue to return different types for multiple outputs (e.g. result_type for data and integers for index). In fact, `result_type` technically only dictates the main execution type, not the actual result type.

**Comment by gchanan** [comment]
Some initial thoughts:
> Omitting same_kind casting from NumPy because torch.dtype.kind isn't a thing.

Why don't we make it a thing?  This is the default for some operations in NumPy (e.g. `numpy.sin`) and IMO sensible behavior (I usually don't care that much about safety, but I don't want to do something dumb like convert a float to an integer implicitly).

> New param casting for torch.as_tensor, with same behavior and default.

Numpy doesn't have this parameter for https://docs.scipy.org/doc/numpy/reference/generated/numpy.asarray.html; this is a little of a tricky situation because I think it makes sense for some contexts (passing a tensor argument and an explicit dtype), but not others (python data, or implicit dtype).  Given that, I'd lean towards not adding this.

> auto (default): Handle casting at dispatch, creating copies of input where necessary
no: casting is not exposed at the API layer (e.g. ones_like)
native: Native function handles casting itself due to one of:

why do we need both 'no' and 'native'?  Why can't a native function just implement 'no'?

>  auto (default): Handle casting at dispatch, creating copies of input where necessary

Why should this be the default?  This seems a bit scary because function immediately changes all behavior.  Are we sure this behavior is sane for all functions?

> Since, for an operation, one backend can have optimized multi-type loops but the others might not, it would be possible to customize by backend:

This doesn't seem like a strong motivation for having backend-specific casting: these cases are pretty rare (and should be non-existing, having different behavior for CUDA vs non-CUDA sucks), and the native function can just implement the difference.

> Native functions will set a casting behavior using the casting attribute.

Why is this a separate attribute rather than a parameter to the function (in the examples above this is a parameter to the function).  The issue here is you are combining two things: 1) the function schema (which presumably has `casting` but it's not listed for some reason)? 2) the function definition

**Comment by tugrulates** [comment]
> Some initial thoughts:
> 
> > Omitting same_kind casting from NumPy because torch.dtype.kind isn't a thing.
> 
> Why don't we make it a thing? This is the default for some operations in NumPy (e.g. `numpy.sin`) and IMO sensible behavior (I usually don't care that much about safety, but I don't want to do something dumb like convert a float to an integer implicitly).
> 
That's valid. It makes sense to me to include as well.

> > New param casting for torch.as_tensor, with same behavior and default.
> 
> Numpy doesn't have this parameter for https://docs.scipy.org/doc/numpy/reference/generated/numpy.asarray.html; this is a little of a tricky situation because I think it makes sense for some contexts (passing a tensor argument and an explicit dtype), but not others (python data, or implicit dtype). Given that, I'd lean towards not adding this.

I had not actually noticed NumPy was missing this. I was thinking about the explicit dtype scenario as well. I don't have a strong argument for adding either.
> 
> > auto (default): Handle casting at dispatch, creating copies of input where necessary
> > no: casting is not exposed at the API layer (e.g. ones_like)
> > native: Native function handles casting itself due to one of:
> 
> why do we need both 'no' and 'native'? Why can't a native function just implement 'no'?

Ah, you're right. There is no need for native.

>
> > auto (default): Handle casting at dispatch, creating copies of input where necessary
> 
> Why should this be the default? This seems a bit scary because function immediately changes all behavior. Are we sure this behavior is sane for all functions?
> 
I think we would ideally arrive at a stage where the default is no-work when adding new functions (which would be whatever implementation auto is). I agree that it should start with the current behavior as default and and functions are covered one-by-one initially.
> > Since, for an operation, one backend can have optimized multi-type loops but the others might not, it would be possible to customize by backend:
> 
> This doesn't seem like a strong motivation for having backend-specific casting: these cases are pretty rare (and should be non-existing, having different behavior for CUDA vs non-CUDA sucks), and the native function can just implement the difference.
> 
The ones I've been looking at recently (cmath wrappers) have all different implementations (THC for cuda vs. ATen/MKL for cpu). I was thinking for flexibility for when merging these, but we can omit this for now.
> > Native functions will set a casting behavior using the casting attribute.
> 
> Why is this a separate attribute rather than a parameter to the function (in the examples above this is a parameter to the function). The issue here is you are combining two things: 1) the function schema (which presumably has `casting` but it's not listed for some reason)? 2) the function definition

You're right the function schema for native casting would change to something like:

```
- func: pow(Tensor self, Tensor exponent, ScalarType dtype=None, Casting casting=None) -> Tensor
```

We would still need something in the schema to pick between default implementation and native implementation. Regarding the above point, something like `autocast_inputs=true|false` could handle that.

**Comment by tugrulates** [comment]
> > > Since, for an operation, one backend can have optimized multi-type loops but the others might not, it would be possible to customize by backend:
> > 
> > 
> > This doesn't seem like a strong motivation for having backend-specific casting: these cases are pretty rare (and should be non-existing, having different behavior for CUDA vs non-CUDA sucks), and the native function can just implement the difference.
> 
> The ones I've been looking at recently (cmath wrappers) have all different implementations (THC for cuda vs. ATen/MKL for cpu). I was thinking for flexibility for when merging these, but we can omit this for now.

Actually, on second thought, customization by backend should be possible to avoid having different behaviors. Not having it would cause behavior deviation.

So a function can:
1. start with autocast for both
2. migrate casting to native for cpu (and specify so in the schema for CPU, but keep CUDA at auto)
3. migrate cuda to native casting

If the schema wouldn't allow this, the behavior between CUDA and CPU would deviate between  steps 2 and 3. You'd be forced to migrate both backends at one go to avoid it.

**Comment by gchanan** [comment]
I guess I'm still not sold on "auto" being particularly useful.  Most functions are structured like:
1) do error checking
2) do wrapping / broadcasting / casting (i.e. the purview of TensorIterator)
3) do calculation

But if you have it done automatically, you have to do it first, which means you spend a bunch of time doing casting before you get to error checking, which isn't great.

Now, this isn't particularly satisfying because moving things to TensorIterator is complicated...but if you have easy ways of doing the casts, you can just implement it one-at-a-time for functions without wholesale moving things to TensorIterator.

Assuming you have that, then you don't need to have backend-specific casting because each backend can just do its casting explicitly and the mechanism doesn't need to get more and more complicated.

**Comment by zou3519** [comment]
#14846 is a good candidate for this

**Comment by tugrulates** [comment]
From above:

> Why don't we use NumPy's behavior?
> NumPy's result_type logic has two undesirable behaviors. The first is that requires examining the actual value of scalars (and 0-dim arrays). This would require a host-device synchronization for 0-dim CUDA tensors. The second is that it often up-promotes 0-dim arrays to float64 or int64.

There is another difference between this and NumPy. NumPy also takes the category (int, float) of scalar into account and uses the scalar type if the scalars have the highest category:

```
>>> 1.5 * torch.tensor(1, dtype=torch.int32)
tensor(1, dtype=torch.int32)
>>> 1.5 * np.array([1], dtype=np.int32)
array([1.5])
```

This is leading to the behavior seen in https://github.com/pytorch/pytorch/issues/15477

I think PyTorch should adopt this behavior for the least surprise. It doesn't lead to either examining the scalar value or promoting float32 tensors to float64 with integer operand.

@colesbury @gchanan any thoughts?

**Comment by gchanan** [comment]
I'm not sure I understand -- doesn't the proposal above cover this via category?

For completeness I turned this into actual code (there appears to be a bug in the above where 0-dim tensors have lower priority than wrapped numbers; I fixed that).  Also note this doesn't handle python ints:
```
import torch

def result_type(*args):
  return torch._promote_types(*tuple(infer_scalar_type(arg) for arg in args if participates(arg, args)))

def infer_scalar_type(arg):
  if isinstance(arg, float):
    return torch.get_default_dtype()
  else:
    return arg.dtype

def participates(arg, args):
  if priority(arg) >= max(priority(other) for other in args):
    return True
  if category(arg) > max(category(other) for other in args if priority(other) > priority(arg)):
   return True
  return False

def priority(arg):
  if isinstance(arg, torch.Tensor) and arg.dim() > 0: return 3
  elif isinstance(arg, torch.Tensor): return 2
  else: return 1

def category(arg):
  if isinstance(arg, float) or arg.dtype.is_floating_point: return 2
  else: return 1

print(result_type(2.5, torch.tensor([[1, 2, 3]])))
```

Result:
```
torch.float32
```

**Comment by tugrulates** [comment]
Ah, yes the proposal covers that but the current implementation in `TensorIterator` doesn't handle it yet. Let me send a PR to add that, which would then fix #15477.

**Comment by dylanbespalko** [comment]
I think what you are proposing here is much more sophisticated than what I am asking for in #17249. While you are discussing dtype promotion, I am also finding that subclassing the torch.Tensor class (as is already done in torch.nn.Parameter) is producing awkward results. For example:

```
p = Parameter(torch.tensor([1, 2, 3]) -> Parameter type
torch.add(1, p) -> **Tensor type**
np.add(1, p) -> Parameter type
```

The numpy implementation of add calls __array_wrap__ even when the desired type is not the first argument in add(). This allows me to cast the result back to Parameter and maintain my sub class type. Please see my example code in #17249. Is it possible to have promotion logic for both dtype and type?

**Comment by mruberry** [comment]
Closing because [torch.result_type](https://pytorch.org/docs/master/generated/torch.result_type.html?highlight=result_type#torch.result_type) is implemented.
